lib/charms/vault_k8s/v0/vault_helpers.py

```diff
@@ -0,0 +1,226 @@
+charm and the vault-k8s charm:
+
+1. Functions that get Vault clients and chain multiple checks on the client.
+2. Functions that chain multiple calls to JujuFacade.
+3. Juju config validation functions.
+4. Utility functions for rendering the Vault config file.
+"""
+
+import logging
+from typing import Dict, List
+
+import hcl
+from charms.vault_k8s.v0.juju_facade import (
+    JujuFacade,
+    NoSuchSecretError,
+)
+from charms.vault_k8s.v0.vault_client import (
+    AppRole,
+    VaultClient,
+    VaultClientError,
+)
+from charms.vault_k8s.v0.vault_managers import AutounsealConfigurationDetails
+from jinja2 import Environment, FileSystemLoader
+from ops.charm import CharmBase
+
+# The unique Charmhub library identifier, never change it
+LIBID = "92129fe159114cf699a24f2e252795a0"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+VAULT_CHARM_APPROLE_SECRET_LABEL = "vault-approle-auth-details"
+CONFIG_TEMPLATE_DIR_PATH = "src/templates/"
+CONFIG_TEMPLATE_NAME = "vault.hcl.j2"
+
+
+class VaultHelpers:
+    """Common functions used by the vault charms."""
+
+    def __init__(self, charm: CharmBase):
+        self.juju_facade = JujuFacade(charm)
+
+    def get_active_and_unsealed_vault_client(
+        self, api_address: str, ca_cert_path: str
+    ) -> VaultClient | None:
+        """Get an active and unsealed Vault client."""
+        if not (vault := self.get_active_vault_client(api_address, ca_cert_path)):
+            return
+        if not vault.is_initialized():
+            logger.error("Can't remove node from cluster - Vault is not initialized")
+            return
+        try:
+            if vault.is_sealed():
+                logger.error("Can't remove node from cluster - Vault is sealed")
+                return
+        except VaultClientError as e:
+            logger.error("Can't remove node from cluster - Vault status check failed: %s", e)
+            return
+        return vault
+
+    def get_unsealed_vault_client(self, api_address: str, ca_cert_path: str) -> VaultClient | None:
+        """Get an unsealed Vault client."""
+        if not (vault := self._get_vault_client(api_address, ca_cert_path)):
+            return
+        if not vault.is_initialized():
+            logger.error("Can't remove node from cluster - Vault is not initialized")
+            return
+        try:
+            if vault.is_sealed():
+                logger.error("Can't remove node from cluster - Vault is sealed")
+                return
+        except VaultClientError as e:
+            logger.error("Can't remove node from cluster - Vault status check failed: %s", e)
+            return
+        return vault
+
+    def get_active_vault_client(self, api_address: str, ca_cert_path: str) -> VaultClient | None:
```

Comment:
Not new in this PR, but I got bit by the naming of this, because it is not necessarily the client for the _active_  vault.

Instead, it is the vault client for the current unit, if it is active or standby (i.e the service is fully initiallized).

In restore backup, we need the client for the  _active_ vault, which you would expect this function to provide you.

If we move forward, we should consider some better naming for these.

Reply:
Moved these parts to #581 and reduced this PR to contain config file related functions only as we discussed.